src/libcore/alloc/mod.rs

```diff
@@ -0,0 +1,333 @@
+//! Memory allocation APIs
+
+#![stable(feature = "alloc_module", since = "1.28.0")]
+
+mod global;
+mod layout;
+
+#[stable(feature = "global_alloc", since = "1.28.0")]
+pub use self::global::GlobalAlloc;
+#[stable(feature = "alloc_layout", since = "1.28.0")]
+pub use self::layout::{Layout, LayoutErr};
+
+use crate::fmt;
+use crate::mem;
+use crate::ptr::{self, NonNull, Unique};
+
+/// The `AllocErr` error indicates an allocation failure
+/// that may be due to resource exhaustion or to
+/// something wrong when combining the given input arguments with this
+/// allocator.
+#[unstable(feature = "allocator_api", issue = "32838")]
+#[derive(Clone, PartialEq, Eq, Debug)]
+pub struct AllocErr;
+
+// (we need this for downstream impl of trait Error)
+#[unstable(feature = "allocator_api", issue = "32838")]
+impl fmt::Display for AllocErr {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("memory allocation failed")
+    }
+}
+
+/// A desired initial state for allocated memory.
+#[derive(Debug, Copy, Clone, PartialEq, Eq)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+pub enum AllocInit {
+    /// The contents of the new memory are undefined.
+    ///
+    /// Reading uninitialized memory is Undefined Behavior; it must be initialized before use.
+    Uninitialized,
+    /// The new memory is guaranteed to be zeroed.
+    Zeroed,
+}
+
+/// Represents a block of allocated memory returned by an allocator.
+#[derive(Debug)]
+#[unstable(feature = "allocator_api", issue = "32838")]
+#[must_use = "`MemoryBlock` should be passed to `AllocRef::dealloc`"]
+pub struct MemoryBlock {
+    ptr: Unique<u8>,
+    layout: Layout,
```

Comment:
That's a fair point. However, when only storing the size, the alignment has to be passed `grow`, `shrink`, and `dealloc`. Do we want to add another parameter like `align: usize` to the mentioned methods then?

Reply:
We could just keep using separate `ptr` and `layout` parameters for these methods, and only use `MemoryBlock` for the returned value. It would then effectively act as a nicer version of `(ptr, new_size)` that was returned in the previous design.

Reply:
We could do it that way, but that would make the reason we introduced `MemoryBlock` somewhat obsolete, since `grow` or `shrink` would return a pointer with `ReallocPlacement::InPlace` (albeit nested in a struct).

Reply:
I don't think that returning a pointer for `InPlace` is such a big issue. Just state in the documentation that this pointer is guaranteed to be the same as the one passed in.